modules/aptx/decode.c

```diff
@@ -0,0 +1,93 @@
+/**
+ * @file aptx/decode.c aptX Decoder
+ *
+ * Copyright (C) 2019 Hessischer Rundfunk
+ */
+
+#include <re.h>
+#include <rem.h>
+#include <baresip.h>
+#include <openaptx.h>
+#include <stdlib.h>
+#include "aptx.h"
+
+struct audec_state
+{
+	struct aptx_context *dec;
+};
+
+static void destructor(void *arg)
+{
+	struct audec_state *ads = arg;
+
+	if (ads->dec)
+		aptx_finish(ads->dec);
+}
+
+int aptx_decode_update(struct audec_state **adsp, const struct aucodec *ac, const char *fmtp)
+{
+	struct audec_state *ads;
+	int err = 0;
+	(void)fmtp;
+
+	if (!adsp || !ac || !ac->ch)
+		return EINVAL;
+
+	ads = *adsp;
+
+	if (ads)
+		return 0;
+
+	ads = mem_zalloc(sizeof(*ads), destructor);
+	if (!ads)
+		return ENOMEM;
+
+	ads->dec = aptx_init(APTX_VARIANT);
+	if (!ads->dec)
+	{
+		warning("Cannot initialize aptX decoder.\n");
+		err = ENOMEM;
+		goto out;
+	}
+
+out:
+	if (err)
+		mem_deref(ads);
+	else
+		*adsp = ads;
+
+	return err;
+}
+
+int aptx_decode_frm(struct audec_state *ads, int fmt, void *sampv, size_t *sampc, const uint8_t *buf, size_t len)
+{
+	size_t processed = 0;
+	size_t written = 0;
+
+	uint8_t *s16 = sampv;
+
+	if (!ads || !sampv || !sampc || !buf)
+		return EINVAL;
+
+	if (fmt != AUFMT_S16LE)
+		return ENOTSUP;
+
+	processed = aptx_decode(ads->dec, buf, len, s16, *sampc, &written);
+
+	*sampc = written / APTX_WORDSIZE;
+
+	if (processed != len)
+		warning("aptX decoding stopped in the middle of the sample, dropped %u bytes\n", (unsigned int)(len - processed));
+
+	if (*sampc > 0)
+	{
+		/* remap S24 to S16 in same buffer */
+		for (size_t s = 0; s < *sampc; s++)
+		{
+			s16[s * 2]     = s16[(s * APTX_WORDSIZE) + 1];
+			s16[s * 2 + 1] = s16[(s * APTX_WORDSIZE) + 2];
+		}
```

Comment:
is the 24-bit format host-endian or little-endian ?

will this code work on big-endian machines ?

we already supports S24_LE and FLOAT, perhaps we should also add those formats here?
(maybe later).

converting from S24 -> S16 means loss of information.

Reply:
libopenaptx documentation says 24-bit little-endian only.
I am not even shure if libopenaptx does work on big-endian machines at all...?

It is only the buffer interface which is always 24-bit wide but most of the time the data is only 16-bit. AptX Standard is 16-bitso the LSB is all zero and this is memory remapping only.
Only aptX HD has support for 24-bit where currently 16-bit/8-bit input will be padded or cropped.

Anyway if it should support native 24_LE I would need some assistance how to implement it.

Maybe L24 module may be also added then ;-)